mobius-core/src/test/java/com/spotify/mobius/MobiusLoopTest.java

```diff
@@ -440,20 +441,51 @@ public void eventsFromEffectHandlerDuringDisposeAreIgnored() throws Exception {
     assertFalse(updateWasCalled.get());
   }
 
+  @Test
+  public void modelsFromInitDuringDisposeAreIgnored() throws Exception {
+    // Model changes emitted from the init function during dispose should be ignored.
+
+    observer = new RecordingModelObserver<>();
+    Semaphore lock = new Semaphore(0);
+
+    final Update<String, TestEvent, TestEffect> update = (model, event) -> Next.noChange();
+    final MobiusLoop.Builder<String, TestEvent, TestEffect> builder =
+        Mobius.loop(update, effectHandler)
+            .init(
+                m -> {
+                  lock.acquireUninterruptibly();
+                  return First.first(m);
+                });
+
+    mobiusLoop = builder.startFrom("foo");
+    mobiusLoop.observe(observer);
+    releaseLockAfterDelay(lock, 30);
+    mobiusLoop.dispose();
```

Comment:
This is racy, too: there are no guarantees that `init` has started executing when `dispose` is called. So there are a few cases here: 

1. `dispose` gets called before `init` starts executing (observer should see nothing).
1. `dispose` gets called while `init` is executing (observer should see nothing).
1. `dispose` gets called after `init` has finished executing but before `dispatchEvent` (observer should see initial model).
1. `dispose` gets called while `dispatchEvent` is running (observer should see initial model).

We should probably try to cover all these cases. It's likely to not be super-easy to write those tests to be completely free from false negatives (that is, errors in production code wouldn't be detected by each test run, but hopefully by enough of them to suffice), because of races between test and production code.

Reply:
Above comment is fixed for this method. AFAICT, it's now race-free and verifies the behaviour when dispose is called before init executes.